Fix compile error on Windows due to invalid constexpr

190n opened the pull request:
#### c3717279a34a046c62d5bd225c79e75c08e390b5
<pre>
Fix compile error on Windows due to invalid constexpr
<a href="https://bugs.webkit.org/show_bug.cgi?id=290001">https://bugs.webkit.org/show_bug.cgi?id=290001</a>

Reviewed by Fujii Hironori.

The functions in Win32HandleMarkableTraits cannot be constexpr because
they read invalidPlatformFileHandle, which is not constexpr, and cannot
be made constexpr because its initializer uses a cast equivalent to
reinterpret_cast.

The Windows definition of invalidPlatformFileHandle is also changed to
use the INVALID_HANDLE_VALUE macro instead of a copy of that macro&apos;s
logic.

* Source/WTF/wtf/FileHandle.h:
(WTF::FileSystemImpl::Win32HandleMarkableTraits::isEmptyValue):
(WTF::FileSystemImpl::Win32HandleMarkableTraits::emptyValue):

Canonical link: <a href="https://commits.webkit.org/292340@main">https://commits.webkit.org/292340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fdb9a1b69af29f7af0bd973f794daa147297dc0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15342 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5257 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100785 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/46239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15630 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23775 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/73006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/46239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98737 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/11701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/86464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53339 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/11426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4191 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/45577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/88402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/4301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102819 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94354 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22792 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23044 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81406 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/25983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15406 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22760 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/27909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/117833 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/22419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/25895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/24161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WTF/wtf/FileHandle.h:
```diff
@@ -44,8 +44,8 @@ const PlatformFileHandle invalidPlatformFileHandle = reinterpret_cast<HANDLE>(-1
-    constexpr static bool isEmptyValue(HANDLE value) { return value == invalidPlatformFileHandle; }
-    constexpr static HANDLE emptyValue() { return invalidPlatformFileHandle; }
+    static bool isEmptyValue(HANDLE value) { return value == invalidPlatformFileHandle; }
+    static HANDLE emptyValue() { return invalidPlatformFileHandle; }
```
Windows has `INVALID_HANDLE_VALUE`. I think we don't need `invalidPlatformFileHandle`. Right?

190n replied:
Not very familiar with the Windows API, but that seems to work too. I'll make the change. The functions still have to be non-constexpr, though:

```
C:\WebKit\Source\WTF\wtf/FileHandle.h(47,27): error: constexpr function never produces a constant expression [-Winvalid-constexpr]
   47 |     constexpr static bool isEmptyValue(HANDLE value) { return value == INVALID_HANDLE_VALUE; }
      |                           ^~~~~~~~~~~~
C:\WebKit\Source\WTF\wtf/FileHandle.h(47,72): note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant expression
   47 |     constexpr static bool isEmptyValue(HANDLE value) { return value == INVALID_HANDLE_VALUE; }
      |                                                                        ^
C:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um\handleapi.h(27,31): note: expanded from macro 'INVALID_HANDLE_VALUE'
   27 | #define INVALID_HANDLE_VALUE ((HANDLE)(LONG_PTR)-1)
      |                               ^
```

190n replied:
Actually, is `invalidPlatformFileHandle` meant to be usable from outside this header? It seems like it would be meant as a cross-platform way to get an invalid handle, but all this code is pretty fresh so nobody outside this header is using `invalidPlatformFileHandle` right now.

I'd at least be inclined to delete it for other OSes too if we're deleting it for Windows. Or, if we want to keep `invalidPlatformFileHandle`, then I could just have the Windows definition be `INVALID_HANDLE_VALUE` instead of the `reinterpret_cast`.

fujii replied:
Got it. Let's keep `invalidPlatformFileHandle`.

fujii reviewed (changes requested):
Please squash commits mannually.

Commit:
Fix compile error on Windows due to invalid constexpr
https://bugs.webkit.org/show_bug.cgi?id=290001

Reviewed by Fujii Hironori.

The functions in Win32HandleMarkableTraits cannot be constexpr because
they read invalidPlatformFileHandle, which is not constexpr, and cannot
be made constexpr because its initializer uses a cast equivalent to
reinterpret_cast.

The Windows definition of invalidPlatformFileHandle is also changed to
use the INVALID_HANDLE_VALUE macro instead of a copy of that macro's
logic.

* Source/WTF/wtf/FileHandle.h:
(WTF::FileSystemImpl::Win32HandleMarkableTraits::isEmptyValue):
(WTF::FileSystemImpl::Win32HandleMarkableTraits::emptyValue):

Canonical link: https://commits.webkit.org/292340@main
